pandas/core/indexes/base.py

```diff
@@ -2654,12 +2658,16 @@ def _union(self, other, sort):
                         stacklevel=3,
                     )
 
-        # for subclasses
-        return self._wrap_setop_result(other, result)
+        return self._shallow_copy(result)
 
     def _wrap_setop_result(self, other, result):
         name = get_op_result_name(self, other)
-        return self._shallow_copy(result, name=name)
+        if isinstance(result, ABCIndexClass):
+            if result.name != name:
```

Comment:
if you want to be really thorough, could handle cases with names like `np.nan` which is not equal to itself.  (but since we're not good about that anyway, OK to leave out-of-scope)

Reply:
Yeah, I saw a note about that in `get_op_result_name` too. I can look into it after this.